### ISSUE TITLE
Unable to send can-fd message with bit-rate switch (BRS)

### ISSUE THREAD

**Issue by walnut-engineering** [deezums/same51_can#3]
First off, sorry I would put this issue in https://github.com/Longan-Labs/CANBed_M4_Arduino_Lib but they are not enabled in that repo.  Perhaps you can help?  @deezums  @stephen1874

I am trying to send and receive can-fd messages with a 1Mbps baud rate, and 2Mbps data rate.  The SAME51_CAN::begin method only took in baud rate and set the data rate to the same value.  I can overloaded the method with a version that takes in data rate separately however the BRS flag is not set according to PCAN-View.  Can you help me with this issue?  I assume there is a register that needs to be set to allow this but I am not intimately familiar with the SAME51 yet..  Thanks in advance!

**Comment by walnut-engineering** [comment]
On further testing it appears the the mode: MCAN_MODE_EXT_LEN_DUAL_RATE is not working.  If I use MCAN_MODE_CAN or MCAN_MODE_EXT_LEN_CONST_RATE I see messages being sent via PCAN-View and the oscilliscope.. however using MCAN_MODE_EXT_LEN_DUAL_RATE I don't see any messages being sent...

**Comment by walnut-engineering** [comment]
Wow, appreciate the quick response!  Managed to get that code to compile and the output was:

> Gclk, Can Baud, NBRP, NTSEG1, NTSEG2, NSJW
> 100000000, 5000, 80, 188, 62, 3, 
> 100000000, 10000, 40, 188, 62, 3, 
> 100000000, 20000, 20, 188, 62, 3, 
> 100000000, 31250, 10, 240, 80, 3, 
> 100000000, 33333, 10, 225, 75, 3, 
> 100000000, 40000, 10, 188, 62, 3, 
> 100000000, 50000, 8, 188, 62, 3, 
> 100000000, 80000, 5, 188, 62, 3, 
> 100000000, 83333, 4, 225, 75, 3, 
> 100000000, 100000, 4, 188, 62, 3, 
> 100000000, 125000, 4, 150, 50, 3, 
> 100000000, 200000, 2, 188, 62, 3, 
> 100000000, 250000, 2, 150, 50, 3, 
> 100000000, 500000, 1, 150, 50, 3, 
> 100000000, 800000, 1, 94, 31, 3, 
> 100000000, 1000000, 1, 75, 25, 3, 
> 
> FD Rates test
> 100000000, 5000, error
> 100000000, 10000, error
> 100000000, 20000, error
> 100000000, 31250, error
> 100000000, 33333, error
> 100000000, 40000, error
> 100000000, 50000, error
> 100000000, 80000, error
> 100000000, 83333, 30, 30, 10, 3, 
> 100000000, 100000, 25, 30, 10, 3, 
> 100000000, 125000, 20, 30, 10, 3, 
> 100000000, 200000, 20, 19, 6, 3, 
> 100000000, 250000, 10, 30, 10, 3, 
> 100000000, 500000, 5, 30, 10, 3, 
> 100000000, 800000, 5, 19, 6, 3, 
> 100000000, 1000000, 4, 19, 6, 3, 
> 100000000, 1250000, 2, 30, 10, 3, 
> 100000000, 1500000, error
> 100000000, 1750000, error
> 100000000, 2000000, 2, 19, 6, 3, 
> 100000000, 2250000, 7, 5, 1, 3, 
> 100000000, 2500000, 1, 30, 10, 3, 
> 100000000, 2750000, 5, 5, 2, 3, 
> 100000000, 3000000, error
> 100000000, 3250000, 7, 3, 1, 3, 
> 100000000, 3500000, error
> 100000000, 3750000, error
> 100000000, 4000000, 1, 19, 6, 3, 
> 100000000, 4250000, 5, 4, 1, 3, 
> 100000000, 4500000, 4, 5, 1, 3, 
> 100000000, 4750000, error
> 100000000, 5000000, 1, 15, 5, 3, 

I'll give these values a shot and report back; thanks again!

**Comment by walnut-engineering** [comment]
Fantastic, that worked!  Now PCAN-View see the frames as valid at 1Mbps baud/2Mbps data rate.  I did need to change GCLK->PCHCTRL[CAN0_GCLK_ID].reg over to use GCLK_PCHCTRL_GEN_GCLK1 and update the system_clock define..

**Comment by ghent360** [comment]
@walnut-engineering can you explain your fix in a bit more details? I'm running into similar issues with the CANBed_M4 board.

**Comment by walnut-engineering** [comment]
@ghent360 from what I recall there were two issues.  First is to change the source clock being used for the CAN is too slow for higher bitrates.  GCLK1 is 48MHz and GCLK2 is 100MHz.  Next update the system_clock define to match.

"same15_can.cpp"
GCLK->PCHCTRL[CAN0_GCLK_ID].reg = GCLK_PCHCTRL_CHEN | GCLK_PCHCTRL_GEN_GCLK**2**;

and 

"mcan.c"
#define __SYSTEM_CLOCK (**100**000000)

**Comment by walnut-engineering** [comment]
Oh, one issue I ran into was editing the files located in the Arduino\libraries folder did not actually change anything and I found that the compiler was using code in ..\AppData\Local\Arduino15\packages\Longan\hardware\samd\1.0.4\libraries\SAME51_CAN\src.. instead.. just fyi

**Comment by ghent360** [comment]
Yeah, I did all that, it is still not working great. The best I got was that it works for 20-30 frames and then stops, because the receiver starts sending errors. (Receiver is Pi4 with MCP2517FD shield)

@deezums does the calculator spit what one should put in the register directly? For example while setting the prescale the code subtracts one from the value: CAN_NBTP_NBRP(regVal32 - 1) is that already computed or I should still subtract 1?

**Comment by walnut-engineering** [comment]
Ah, yes forgot to mention that I did need to update the values that we're spit out by @deezums application..  if you look at the src.zip I attached in previous post.  The existing library also did not support different arbitration rates and data rates so I added that.

**Comment by ghent360** [comment]
I finally got it working. I made @deezums calculator work on a desktop PC: https://gist.github.com/ghent360/fe34f4ec663db133117e92e70445cc6f

The values go straight into the config structure, no need to subtract or add anything.

In my case I was using the latest library from Longan's SAME51 Arduino package, I'm not sure the core is public. They added some hack in the code with a global variable called fd_mode and it was not set correctly. That is why it was not initializing the fd timing registers at all and I was confused why me changing the quanta values had no effect.

I made a PR to the Longan's can lib depot with the changes https://github.com/Longan-Labs/CANBed_M4_Arduino_Lib/pull/2 thins includes the diffs between that repo and the code that was downloaded as part of the arduino BSP.

Tested speeds up to 1M_5M and up to 12 hours w/o problems.

Thanks @deezums and @walnut-engineering  for the help.
